### ISSUE TITLE
Proposal: zaptar rune for statically typed compilation and evaluation

### ISSUE THREAD

**Issue by belisarius222** [urbit/urbit#1581]
@benjamin-tlon @pilfer-pandex @philipcmonk @joemfb @eglaysher

I propose a new rune for typed runtime compilation and evaluation. This is not obviously possible in a statically typed language without dependent types, but due to Hoon's compilation to Nock, we can do it safely and without performance penalty. For an intriguing (and much more complicated) analog in System F, see [Self-Quotation in a Typed, Intensional Lambda-Calculus](https://coalg.org/mfps-calco2017/mfps-papers/4-jay.pdf).

This does have to be done through a rune to get the static proof of type safety, and it requires adding a new biblical. I have a partial implementation, but thought I'd present it early on to get feedback before putting real work into it.

The rune compiles a Hoon AST passed in at runtime against a statically specified subject type and result type, then evaluates the resulting Nock using a subject specified at runtime.

The syntax looks like this:
```
!*  result-example  subject  some-hoon
```

If the compilation fails, the expression crashes. Deterministic crashes can be caught by virtualizing with `+mule`, as usual.

If the compilation succeeds, we have a new value whose type is statically known, even though it was compiled at runtime. This would remove the need for vases in several parts of the system, which I believe will spark joy.

I've written a toy example of how Gall could use this if we had a static type for apps. The Arvo kernel could also use this if vanes had a static type. As @frodwith suggested, `+eve` itself could be jetted as a way for the runtime to make use of Hoon types for optimization. The runtime could also memoize it, replacing the Ford cache. This rune does not require any runtime support, though.

Note the conspicuous absence of vases:
```
::  |app-runner: like Gall, but statically typed
::
=>  |%
    +$  app-event   [%foo @]
    +$  app-effect  [%bar ^]
    +$  app-core
      $_  ^?
      |%
      ++  poke  |=(app-event [(list app-effect) app-core])
      ++  peek  |=(path (unit (unit [mark *])))
      --
    --
::
|_  apps=(map @tas app-core)
++  app-runner  .
::
++  poke-app
  |=  [app-name=@tas =app-event]
  ^-  [(list app-effect) _app-runner]
  ::
  =/  app=app-core  (~(got by apps) app-name)
  =^  effects  app  (poke:app app-event)
  =.  apps          (~(put by apps) app-name app)
  ::
  [effects app-runner]
::
++  peek-app
  |=  [app-name=@tas =path]
  ^-  (unit (unit [mark *]))
  ::
  =/  app=app-core  (~(got by apps) app-name)
  (peek:app path)
::
++  install-app
  |=  [app-name=@tas source=@t]
  ^+  app-runner
  ::
  ::  compile the source to an +app-core, catching failure with +mule
  ::
  =/  eval-result=(each app-core tang)
    %-  mule  |.
    ~&  %compiling-app^app-name
    ::
    !*  *app-core  ..zuse  (ream source)  ::  <----  where the magic happens
  ::
  ::  print and no-op on failure
  ::
  ?:  ?=(%| -.eval-result)
    %-  (slog p.eval-result)
    app-runner
  ::
  ::  install the new app in our state
  ::
  =.  apps  (~(put by apps) app-name p.eval-result)
  ::
  app-runner
--
```

As for implementation, the rune's type signature looks like this:
```
[%zptr result=hoon subject=hoon =hoon]
```
`:subject` should evaluate to the subject against which the resulting Nock should be evaluated. Its type will be analyzed statically and used as an argument to `+mint:ut` at runtime. `:result` is an example result whose type will be extracted at compile time and passed into the `+mint:ut` runtime call to constrain the output type of the compilation. `:hoon` is an expression that should produce an AST to be compiled and run against `:subject`.

This ends up calling code that looks like this at runtime (`+eve` is the new biblical):
```
++  eve
  |=  [=result=type subject=[=type value=*] =hoon]
  ^-  *
  ::
  .*  value.subject
  +:(~(mint ut type.subject) result-type hoon)
```

The compilation of the rune in `+mint:ut` looks something like this:
```
  ?-  gen
  ...
        [%zptr *]
      ^-  [type nock]
      ::  obtain type from :result.gen, ignoring type from rest of expression
      ::
      :-  (play result.gen)
      ::  produce nock that runs the compiler using +eve
      ::
      =<  +
      %_    $
          gen
        ^-  hoon
        :+  %cncl  [%limb %eve]
        :~  [%tsld [%$ 2] %zpbn result.gen]
            [%zpbn subject.gen]
            hoon.gen
        ==
      ==
  ...
  ==
```

**Comment by ohAitch** [comment]
I am mildly concerned that this would make language revisions considerably
more difficult, and would request an OTA update that makes a trivial change
to ++twig in a system using this rune be done as part of
testing/validation; but this is definitely the kind of static
metaprogramming capacity that is aesthetically squarely in Urbit's domain!

I have previously thought at some length about the "statically typed arvo"
style this enables, though based on a simpler vase->value lifter; it's
definitely cleaner than having explicit vase compiles and calls all over
the place.

**Comment by belisarius222** [comment]
Yeah, the vase->value lifter is the !< rune that Philip just proposed.  It certainly supersedes this rune proposal.

—
~rovnys-ricfer
https://urbit.org
